src/ResourceManager/Common/Commands.ResourceManager.Common/RPRegistrationDelegatingHandler.cs

```diff
@@ -0,0 +1,119 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Commands.ResourceManager.Common.Properties;
+using Microsoft.Azure.Common.Authentication;
+using Microsoft.Azure.Common.Authentication.Models;
+using Microsoft.Azure.Management.Resources;
+using Microsoft.Azure.Management.Resources.Models;
+using Microsoft.WindowsAzure.Commands.Utilities.Common;
+
+namespace Microsoft.Azure.Common.Authentication.Models
+{
+    public class RPRegistrationDelegatingHandler : DelegatingHandler, ICloneable
+    {
+        private const short RetryCount = 3;
+
+        /// <summary>
+        /// Contains all providers we have attempted to register 
+        /// </summary>
+        private HashSet<string> registeredProviders;
+
+        private Func<ResourceManagementClient> createClient;
+
+        private Action<string> writeDebug;
+
+        public ResourceManagementClient ResourceManagementClient { get; set; }
+
+        public RPRegistrationDelegatingHandler(Func<ResourceManagementClient> createClient, Action<string> writeDebug)
+        {
+            registeredProviders = new HashSet<string>();
+            this.writeDebug = writeDebug;
+            this.createClient = createClient;
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            ResourceManagementClient = createClient();
+
+            HttpResponseMessage responseMessage = await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
+            if (IsProviderNotRegistereError(responseMessage))
+            {
+                var providerName = GetProviderName(request.RequestUri);
+                if (!string.IsNullOrEmpty(providerName) && !registeredProviders.Contains(providerName))
+                {
+                    // Force dispose for response messages to reclaim the used socket connection.
+                    responseMessage.Dispose();
+                    registeredProviders.Add(providerName);
+                    try
+                    {
+                        writeDebug(string.Format(Resources.ResourceProviderRegisterAttempt, providerName));
+                        ResourceManagementClient.Providers.Register(providerName);
+                        Provider provider = null;
+                        short retryCount = 0;
+                        do
+                        {
+                            if (retryCount++ > RetryCount)
+                            {
+                                throw new TimeoutException();
```

Comment:
Add in a message here so the timeout shows up in the exception-catching trace below (or add a specific trace for the timeout)